client-data/tools/document/document.js

```diff
@@ -0,0 +1,103 @@
+(function documents() { //Code isolation
+
+
+    var xlinkNS = "http://www.w3.org/1999/xlink";
+    var imgCount = 1;
+
+    function assert_count() {
+        if (Tools.svg.querySelectorAll("image").length >= Tools.server_config.MAX_DOCUMENT_COUNT) {
+            alert("Too many documents exist already");
+            throw new Error("Too many documents exist already");
+        }
+    }
+
+    function onstart() {
+        var fileInput = document.createElement("input");
+        fileInput.type = "file";
+        fileInput.accept = "image/*";
+        fileInput.click();
+        fileInput.addEventListener("change", function () {
+            assert_count();
+
+            var reader = new FileReader();
+            reader.readAsDataURL(fileInput.files[0]);
+
+            reader.onload = function (e) {
+                // use canvas to compress image
+                var image = new Image();
+                image.src = e.target.result;
+                image.onload = function () {
+
+                    assert_count();
+
+                    var uid = Tools.generateUID("doc"); // doc for document
+
+                    var ctx, size;
+                    var scale = 1;
+
+                    do {
+                        // Todo give feedback of processing effort
+
+                        ctx = document.createElement("canvas").getContext("2d");
+                        ctx.canvas.width = image.width * scale;
+                        ctx.canvas.height = image.height * scale;
+                        ctx.drawImage(image, 0, 0, image.width * scale, image.height * scale);
+                        var dataURL = ctx.canvas.toDataURL("image/webp", 0.7);
+
+                        // Compressed file size as data url, approximately 1/3 larger than as bytestream
+                        size = dataURL.length;
+
+                        // attempt again with an image that is at least 10% smaller
+                        scale = scale * Math.sqrt(Math.min(
+                                0.9,
+                                Tools.server_config.MAX_DOCUMENT_SIZE / size
+                        ));
+                    } while (size > Tools.server_config.MAX_DOCUMENT_SIZE);
+
+                    var msg = {
+                        id: uid,
+                        type: "doc",
+                        data: dataURL,
+                        size: size,
+                        w: this.width * scale || 300,
```

Comment:
This uses the newly reduced value for `scale`. I think it was meant to use the previous value, which is the value used with `drawImage()`